📖 : Proposal : Discontinue Kube RBAC Proxy in Default Kubebuilder Scaffolding (#3860)
kubernetes-sigs/kubebuilder: designs/discontinue_usage_of_kube_rbac_proxy.md

These enhancements would introduce encrypted communication for metrics endpoints and potentially incorporate authentication mechanisms,
significantly elevating the security model employed by projects scaffolded by Kubebuilder.

- **cert-manager**: Automates the management and issuance of TLS certificates, facilitating encrypted communication and, when configured with mTLS, adding a layer of authentication.
  Currently, we leverage on cert-manager when webhooks are scaffold. So, the proposal idea would be to allow users enable the cert-manager for the metrics such as it is provided
  and required for webhook feature. However, it MUST be optional. One of the goals of Kubebuilder is make it easier for new users, therefore new users should
  not need to deal with cert-manager by default or have the need to install it to just quick start.

- **Controller-Runtime Feature (PR #2407)[cr-pr]**:  introduces options for metrics that allow serving metrics via TLS and using authentication and authorization.
It seems fit very well to our requirements.

While this setup does not replicate [kube-rbac-proxy's][kube-rbac-proxy] method of leveraging Kubernetes RBAC for authentication

> and authorization directly, it achieves a comparable level of security through a combination of network isolation and encrypted communication.

Reviewer: I think this setup _does_ do the same thing as kube-rbac-proxy. See https://github.com/kubernetes-sigs/controller-runtime/blob/cf69be219f050a456a93949d0c53773836f9216c/pkg/metrics/filters/filters.go#L29-L32

Reviewer 2: It does!

Author: Closing since it is addressed in the proposal as well.

the patches in the `config/default/kustomization` and disable the default network policy used within
and [replace the code](https://github.com/kubernetes-sigs/kubebuilder/blob/72586d386cfbcaecea6321a703d1d7560c521885/pkg/plugin/util/util.go#L231)
in the `main.go` bellow with in order to not use the controller-runtime
feature instead.

```go
    ctrlOptions := ctrl.Options{
		MetricsFilterProvider: filters.WithAuthenticationAndAuthorization,
		MetricsSecureServing:  true,
	}
```

### Proof of Concept

- **(Phase 1)NetworkPolicies:** https://github.com/kubernetes-sigs/kubebuilder/pull/3853
- Example of Controller-Runtime new feature to protect Metrics Endpoint: https://github.com/sbueringer/controller-runtime-tests/tree/master/metrics-auth

### Risks and Mitigations

#### Loss of Previously Promoted Images

The transition to the new shared infrastructure for Kubernetes SIG projects has rendered us unable to automatically
build and promote images as before. The process only works for projects under the umbrella.
However, the k8s-infra maintainers are assisting in manually transferring these images

> to the new [registry.k8s.io][registry.k8s.io].

Author: The images are manually added there now; https://explore.ggcr.dev/?repo=gcr.io%2Fk8s-staging-kubebuilder%2Fkube-rbac-proxy

Author: added so closing

kube-rbac-proxy seems to be to direct them to the images hosted
at [quay.io](https://quay.io/repository/brancz/kube-rbac-proxy?tab=tags&tag=latest),
which are maintained by the project itself and then,
we keep those images in the registry.k8s.io as a "contingent approach".

Ensuring that these images will continue to be promoted under any infrastructure available to
Kubebuilder is not reliable or achievable for Kubebuilder maintainers. It is definitely out of our control.

#### Challenges in Enhancing Metrics Endpoint Protection

Our initial phase of implementing [Network Policies][k8s-doc-networkpolicies] provides a basic level of protection.
The combination with cert-manager, while not yet proven, is anticipated to be a reliable and common practice
for securing communications.

Should integration with controller-runtime prove challenging or if such functionality
is removed from the project, our stance is to offer the best feasible security measures
that we can confidently support, the changes done so far already provided security and
are aligned with Kubernetes common practices and most important, are solutions

> that we can properly support.

Author: Hi @sbueringer,

You are the author of the controller-runtime feature that we are thinking in using here + cert-manager and networkpolicies (https://github.com/kubernetes-sigs/controller-runtime/pull/2407)

I did not POC that and that is the reason of I add this section.
However, I would like to ask a hand for you. If you do not see any reason for that does not work, I think we can remove this part and make this doc less verbose. 

Have you any thoughts on that?

Reviewer 2: I think the controller-runtime feature works well to protect the metrics endpoint.

We use it in CAPI since a while now.

Here is basically the configuration you need: https://github.com/kubernetes-sigs/cluster-api/blob/500fc1fec6867a5b27b7bda38882f7d1957a8b52/util/flags/diagnostics.go#L79-L82
(the metric options can be set via ctrl.Options / Metrics)

Per default the metrics server will use a self-signed certificate, but you can also configure it with CertDir / .. option to use the same certificate as the webhooks server.

Author: Hi @sbueringer 

Thank you a lot for your info.

Regards a self-signed certificate:

I understand that one of the breaking changes of kube-rbac-proxy is to remove that. The reason behind that is that after the auth-sig reviews this option seems not safe :-( So, if that is right we cannot use it and we must config with cert-manager. 

But anyway, it is great to know that I can remove this section. 
Doing it now.

> We tried to do that, see [here](https://github.com/kubernetes/test-infra/blob/master/config/jobs/image-pushing/k8s-staging-kubebuilder.yaml) the recipe implemented.
> However, it does not work because kube-rbac-proxy is not under the
> kubernetes umbrella. More over we experiment the GitHub Repository as alternative approach, see the [PR](https://github.com/kubernetes-sigs/kubebuilder/pull/3854) but seems
> that we are not allowed to use it. Nevertheless, both approaches does not show to sort out all motivations and requirements
> Ideally, Kubebuilder should not be responsible for maintain and promote third-part artifacts.

1. However, is not Kubebuilder also building and promoting the binaries required to be used within [EnvTest](./../docs/book/src/reference/envtest.md)
feature implemented in controller-runtime?

> Yes, but it also will need to change. Controller-runtime maintainers are looking for solutions to
> built those binaries inside its project since it seems part of its domain. This change is likely
> to be transparent to the community users.

> 1. Could we not use the Controller-Runtime feature [controller-runtime][cr-pr] which enable secure metrics serving over HTTPS?

Author: That is a great idea. !!!!

I will add the option of we enhance in a follow up when controller-runtime be updated.
@sbueringer since you are the author of the feature you might want to give a look on that.

If we be able to improve that we can add it in the Kubebuilder in the future for sure
that would be a great option

> 1. Could we not use the Controller-Runtime feature [controller-runtime][cr-pr] which enable secure metrics serving over HTTPS?

Author: Closing since we updated the proposal within.

No. After we ask for a hand for reviews from skilled auth maintainers and receiving feedback, it appears that this configuration does not
align with best practices. Currently, options such as cache TTL, anonymity, and timeouts are not configurable,
and there is no way to override them. Moreover, the certificate authentication seems to be broken.
The `alwaysallow` option for certain paths (like `healthz`) is not configurable, and the `alwaysallowgroups` (e.g., `system:masters`)
does not seem to be properly configured. This configuration is likely to cause issues with vendors like OpenShift,
which provide Prometheus for the entire cluster. See how this vendor attempts to overwrite some configurations to

> work properly within their environment: [OpenShift kube-rbac-proxy pull request #43](https://github.com/openshift/kube-rbac-proxy/pull/43/files).

Reviewer 3: not openshift specific.  See https://github.com/brancz/kube-rbac-proxy/pull/125.

Author: I will update it here. Thank you for the help.

Author: Closed since it is addressed in the proposal.
We are linking the issue open to Controller-Runtime

the Kubernetes cluster. While support and implementation details can vary among CNIs, it seems that the most commonly used ones,
such as Calico, Cilium, WeaveNet, and Canal, shows to offer  support for NetworkPolicies.
>
>Also, there was concern in the past because AWS did not support it. However, this changed,
>as detailed in their announcement: [Amazon VPC CNI now supports Kubernetes Network Policies](https://aws.amazon.com/blogs/containers/amazon-vpc-cni-now-supports-kubernetes-network-policies/).
>
>Moreover, under this proposal users still able to disable/enable this option as please them.

1. NetworkPolicy is a simple firewall and does not provide authn/authz and encryption.

> Yes, that's correct. NetworkPolicy acts as a basic firewall for pods within a Kubernetes cluster, controlling traffic
> flow at the IP address or port level. However, it doesn't handle authentication (authn), authorization (authz),
> or encryption directly like kube-rbac-proxy solution.
>
> However, if we be able to combine the cert-manager and the new feature provided
> by controller-runtime we can achieve the same or a superior level of protection

> > without relay in any extra third-party dependency.

Reviewer 4: ```suggestion
> without relying on any extra third-party dependency.
```

> > without relay in any extra third-party dependency.

Reviewer 4: link to where this is documented?

Author: Hi @joejulian,

Thank you for your review and time.
Could you please let us know what documentation are you looking for here?
What exactly clarifications do you need in the OpenQuestion answered.

Reviewer 4: Someone who was previously relying on kube-rbac-proxy wants to now follow the new recommended method of using cert-manager and the new mtls feature (if I'm understanding this correctly), where do they find the documentation for that? Could we link that here?

Author: That is the proposal for change 
We are proposing it all get done in phases
So, the first phase does not use cert-manager 
See: https://github.com/kubernetes-sigs/kubebuilder/pull/3853

Then, a follow up of this PR we need to add a config in the kustomize
that will allow we patch the serviceMonitor and certs such as we do now for webhooks 
See that we have in the config/default/ a patch for cert-manager + webhooks: 
See https://github.com/kubernetes-sigs/kubebuilder/blob/bd0876b8132ff66da12d8d8a0fdc701fde00f54b/docs/book/src/component-config-tutorial/testdata/project/config/default/kustomization.yaml#L51-L149 and https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v4/config/default/manager_webhook_patch.yaml

In the follow up we will need to do something similar 
We will need to have something like

```
# [METRICS WITH HTTPS] To enable the ServiceMonitor using HTTPS uncomment the following line
# Note that for that works you also need ensure that you enable cert-manager in your project
- path: metrics_https_patch.yaml
```

That should patch https://github.com/kubernetes-sigs/kubebuilder/pull/3853/files#diff-8545904ae77cfcef7da403b4fe25b822439217abe9307d5d31b29690140d0fc6

As part of the task we will need also to update the docs to clarify how that works and etc.

Author: Hi @joejulian 

It was also added in the proposal.
See that we are adding examples either how the YAML would end up.

Thank you for raise it out.

1. Could not Kubebuilder maintainers use the shared infrastructure to continue building and promoting those images under the new `registry.k8s.io`?

> We tried to do that, see [here](https://github.com/kubernetes/test-infra/blob/master/config/jobs/image-pushing/k8s-staging-kubebuilder.yaml) the recipe implemented.
> However, it does not work because kube-rbac-proxy is not under the
> kubernetes umbrella. More over we experiment the GitHub Repository as alternative approach, see the [PR](https://github.com/kubernetes-sigs/kubebuilder/pull/3854) but seems

> > that we are not allowed to use it. Nevertheless, both approaches does not show to sort out all motivations and requirements

Reviewer 4: > ... both approaches does not show to sort out all ...
I think there's a typo here, but I'm not sure what this is trying to say.

Author: It is a section of questions and answers
We are trying to clarify here that:
- We cannot use the e2e shared infra and why
- We cannot use the github packages/repository as well

Anyway, if we would be able to address the needs within any of those approaches that does NOT sort out ALL our problems and motivations. Either does not seems the best approach for the long term.

However, please feel free to let us know what is not clear and provide any suggestion.

Reviewer 4: Got it. How about

```suggestion
> that we are not allowed to use it. Nevertheless, neither approach sorts out all motivations and requirements
```

These enhancements would introduce encrypted communication for metrics endpoints and potentially incorporate authentication mechanisms,
significantly elevating the security model employed by projects scaffolded by Kubebuilder.

- **cert-manager**: Automates the management and issuance of TLS certificates, facilitating encrypted communication and, when configured with mTLS, adding a layer of authentication.
  Currently, we leverage on cert-manager when webhooks are scaffold. So, the proposal idea would be to allow users enable the cert-manager for the metrics such as it is provided
  and required for webhook feature. However, it MUST be optional. One of the goals of Kubebuilder is make it easier for new users, therefore new users should
  not need to deal with cert-manager by default or have the need to install it to just quick start.

### Phase 3: When Controller-Runtime feature be enhanced

After we have the [issue](https://github.com/kubernetes-sigs/controller-runtime/issues/2781)
addressed we plan to use it to protect the endpoint. See that would mean ensure
that we are either `handle authentication (authn), authorization (authz)`.

> Examples of its implementation can be found [here](https://github.com/sbueringer/controller-runtime-tests/blob/master/metrics-auth/withmetricsauth/main.go#L10-L14)

Reviewer 2: I would prefer linking to the CAPI implementation instead: https://github.com/kubernetes-sigs/cluster-api/blob/v1.6.3/util/flags/diagnostics.go#L79-L82

Reviewer 2: (also the second link below)

Author: ```suggestion
Examples of its implementation can be found [here](https://github.com/kubernetes-sigs/cluster-api/blob/v1.6.3/util/flags/diagnostics.go#L79-L82) 
```

### Phase 4: When kube-rbac-proxy be accepted under the umbrella

Once kube-rbac-proxy is included in the Kubernetes umbrella,

> Kubebuilder maintainers can support its integration through a [plugin](https://kubebuilder.io/plugins/plugins).

Reviewer 5: Can we also have the link to the issue that tracks the inclusion of kube-rbac-proxy to k8s-apimachinery umbrella ?

Author: Done
